### ISSUE TITLE
A request for adding new localizations(lang and manual).

### ISSUE THREAD

**Issue by MerryYukari** [ReikaKalseki/Reika_Mods_Issues#1824]
Dear reika, apology for bothering you, I've completed translation of [this repository](https://github.com/MerryYukari/Reika_mods_Localization), including Chromaticraft, RotaryCraft, ReactorCraft and ElectriCraft's lang files and handbooks for V19c, also, test in local game passed except RotaryCraft's handbook because new translation files failed security check. So I want to ask if you could implement translation files in new version of these mods.
My best wishes.

*I'm a completely newbie to github, and I don't know if I should pull request first or something like reopen #155 and #211, however the [original repository](https://github.com/UUUii/Reika_mods_Localization) seems like abandoned for a while.

**Comment by ReikaKalseki** [comment]
If you have new localized files that you want added, just link them and I can add them to the project on my end.

**Comment by MerryYukari** [comment]
Sure, links are here:
RotaryCraft: [lang file](https://github.com/MerryYukari/Reika_mods_Localization/blob/master/Rotarycraft/zh_CN.lang), [handbook](https://github.com/MerryYukari/Reika_mods_Localization/tree/master/Rotarycraft/Resources/zh_CN);
ReactorCraft: [lang file](https://github.com/MerryYukari/Reika_mods_Localization/blob/master/ReactorCraft/zh_CN.lang), [handbook](https://github.com/MerryYukari/Reika_mods_Localization/tree/master/ReactorCraft/Resources/zh_CN);
ElectriCraft: [lang file](https://github.com/MerryYukari/Reika_mods_Localization/blob/master/ElectriCraft/zh_CN.lang), [handbook](https://github.com/MerryYukari/Reika_mods_Localization/tree/master/ElectriCraft/Resources/zh_CN);
ChromatiCraft: [lang file](https://github.com/MerryYukari/Reika_mods_Localization/blob/master/ChromatiCraft/zh_CN.lang), [lexicon](https://github.com/MerryYukari/Reika_mods_Localization/tree/master/ChromatiCraft/Resources/zh_CN).
My best wishes.

*Excuse me, I saw many text in Chroma Lexicon just get outside of the book's boundary, this also happened to the side bar while looking at a ChromatiCraft block, I do see issues like this, but is there any possibility to fix? Or can I just manually add newlines to avoid this?

**Comment by ReikaKalseki** [comment]
I have no idea how Chinese script works, so I know nothing about how it might handle newlines, either computationally or how it might affect meaning.

**Comment by MerryYukari** [comment]
So, I guess it's best to keep it just what it looks like currently...
My best wishes.

**Comment by 3TUSK** [comment]
Apologize for jumping in. [As the past main maintainer of that repository](https://github.com/MerryYukari/Reika_mods_Localization/graphs/contributors), I have some concerns on this request:

1. Is there any protection, or fallback mechanism (reads "display English when in error"), that will prevent game from crashing due to malformated localization files, if such error accidentally occurs (for example #517)? This is the major reason of that I decided to hold on and not requesting on adding new localization to ChromatiCraft back to mid-2016 when I initially finished translation of Chroma Lexicon - I don't know if any of my translation would be invalid in the future, and I cannot guarantee that I have enough time to keep them up-to-date. (And, as far as in my knowledge, none of past maintainers of that repositories has time to keep localization up-to-date.)
2. Not really on-topic, but - is there any chance that you can set up a repository, dedicating on collecting localizations, for your mods?
3. Regarding line breaking in Chinese script. As Chinese script does not use whitespace as word delimiter, the [word wrapping rules for Chinese script](https://en.wikipedia.org/wiki/Line_wrap_and_word_wrap#Word_wrapping_in_text_containing_Chinese,_Japanese,_and_Korean) differs from that of Latin script. `java.text.BreakIterator` can help determine proper word wrapping place in language-agnostic manner. The only question is whether adapting `BreakIterator` to your codebase is practical or not. 
4. This question is for @MerryYukari. 
Are you able to keep updating those localizations in foreseeable future? 在可以预见的未来中，你能不能保证所有相关的本地化（RotaryCraft、ReactorCraft、ElectricCraft 和 ChromatiCraft）的及时更新？

**Comment by MerryYukari** [comment]
Yeah, I do understand 3TUSK's concerns. Also, there're some points that I've noticed during localizing language files.

1. As for all mods mentioned above, using CRLF line endings for xml files won't crash Minecraft, but the game won't load localized files too;
2. For ChromatiCraft, 'categories.xml' is needed for localization to work, although that file doesn't contain any usable strings(I guess);
3. For ChromatiCraft, in the Chinese environment, text in Lexicon entries and side bars will only warp at the point of Latin punctuation;
4. Replacing zh_CN.lang may cause missing blocks, but I'm not sure since that only happened once during my many tests for translation files;

Reply to @3TUSK .
As for keep updating localizations, as long as I'm still playing Minecraft, I'll keep it up-to-date, and this should be a quite long time. However, it looks like new blocks and entries will be added in new numbered releases like v20a, and I need some time to translate new entries and proofread old entries, so what I can do is to hope versions like v20a won't crash, but some entries of newly added blocks in old RotaryCraft handbook just display as #NULL, so I guess that should be fine.
只要我还在玩Minecraft，我应该就会继续翻译。但是翻译工作也需要一些时间，所以不能确定是否及时。而且在旧版的RotaryCraft手册里一些新添加的方块的说明是#NULL（没有汉化条目），所以我猜新方块应该不会崩掉游戏。

My best wishes.

**Comment by sfmiyuki** [comment]
To MerryYukari: @MerryYukari.
If you miss entries in CC, it will crash(in few versions ago, not sure now but possibly), which means at least that you MUST update CC's localization asap. Otherwise all Chinese will have no chance to use it.

如果ChromatiCraft汉化手册里面漏掉了条目就会崩（几个版本前的事，但我觉得他应该没改），也就是说如果你更新脱节一个版本的话意味着中文再也无法使用

**Comment by ReikaKalseki** [comment]
@3TUSK I implemented a fallback-to-English-if-localized-files are broken system. I also created a repository for non-handbook localization data:

https://github.com/ReikaKalseki/Localizations

**Comment by MerryYukari** [comment]
Thanks for the mention, I just finished translation for V20, but what should I do now, fork this localization repository first?
Also, is anything I can do about those translated handbook?

My best wishes.

**Comment by ReikaKalseki** [comment]
I am not sure what you are asking.
